pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -529,36 +523,49 @@ func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
 	return svc, nil
 }
 
+// getNodePortServices returns nodePort type gateway service
+func (c *Controller) getNodePortGatewayServices() []*model.Service {
+	c.RLock()
+	defer c.RUnlock()
+	out := make([]*model.Service, 0, len(c.nodeSelectorsForServices))
+	for hostname := range c.nodeSelectorsForServices {
+		svc := c.servicesMap[hostname]
+		if svc != nil {
+			out = append(out, svc)
+		}
+	}
+
+	return out
+}
+
 // updateServiceExternalAddr updates ClusterExternalAddresses for ingress gateway service of nodePort type
 func (c *Controller) updateServiceExternalAddr(svcs ...*model.Service) {
 	// node event, update all nodePort gateway services
 	if len(svcs) == 0 {
-		svcs, _ = c.Services()
+		svcs = c.getNodePortGatewayServices()
 	}
 	for _, svc := range svcs {
-		if isNodePortGatewayService(svc) {
-			c.RLock()
-			nodeSelector := c.nodeSelectorsForServices[svc.Hostname]
-			c.RUnlock()
-			// update external address
-			svc.Mutex.Lock()
-			if nodeSelector == nil {
-				var extAddresses []string
-				for _, n := range c.nodeInfoMap {
-					extAddresses = append(extAddresses, n.address)
-				}
-				svc.Attributes.ClusterExternalAddresses = map[string][]string{c.clusterID: extAddresses}
-			} else {
-				var nodeAddresses []string
-				for _, n := range c.nodeInfoMap {
-					if nodeSelector.SubsetOf(n.labels) {
-						nodeAddresses = append(nodeAddresses, n.address)
-					}
+		c.RLock()
+		nodeSelector := c.nodeSelectorsForServices[svc.Hostname]
+		c.RUnlock()
+		// update external address
+		svc.Mutex.Lock()
+		if nodeSelector == nil {
+			var extAddresses []string
+			for _, n := range c.nodeInfoMap {
+				extAddresses = append(extAddresses, n.address)
+			}
+			svc.Attributes.ClusterExternalAddresses = map[string][]string{c.clusterID: extAddresses}
+		} else {
+			var nodeAddresses []string
+			for _, n := range c.nodeInfoMap {
+				if nodeSelector.SubsetOf(n.labels) {
+					nodeAddresses = append(nodeAddresses, n.address)
```

Comment:
If we collect the node ips from the pods, all this logic goes away.

Reply:
yes